cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,538 @@
+    if (action.balancingAction() != ActionType.INTER_BROKER_REPLICA_SWAP) {
+      throw new IllegalArgumentException("Expected INTER_BROKER_REPLICA_SWAP. But got: " + action.balancingAction());
+    }
+    Replica sourceReplica = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
+    Replica destinationReplica = clusterModel.broker(action.destinationBrokerId()).replica(action.topicPartition());
+    if (sourceReplica.isLeader() == destinationReplica.isLeader() && Objects.equals(action.topic(), action.destinationTopic())) {
+      // Swapping either 2 leader replicas or 2 follower replicas of the same topic has no effect on this goal
+      return true;
+    }
+    return isSelfSatisfiedToMoveOneLeaderReplica(clusterModel.broker(action.sourceBrokerId()),
+                                                 clusterModel.broker(action.destinationBrokerId()),
+                                                 sourceReplica)
+           && isSelfSatisfiedToMoveOneLeaderReplica(clusterModel.broker(action.destinationBrokerId()),
+                                                    clusterModel.broker(action.sourceBrokerId()),
+                                                    destinationReplica);
+  }
+
+  private boolean actionAffectsRelevantTopics(BalancingAction action) {
+    if (_mustHaveTopicLeadersPerBroker.contains(action.topic())) {
+      return true;
+    }
+    return action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP &&
+           _mustHaveTopicLeadersPerBroker.contains(action.destinationTopic());
+  }
+
+  /**
+   * Check whether it is self-satisfied to move a leader replica of a topic with the given name from a the given source
+   * broker to the given destination broker
+   * @param sourceBroker the broker from which a leader replica is removed
+   * @param destinationBroker the broker from which a leader replica is added
+   * @param replicaToBeMoved replica to be moved from the source broker to the destination broker
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean isSelfSatisfiedToMoveOneLeaderReplica(Broker sourceBroker, Broker destinationBroker, Replica replicaToBeMoved) {
+    if (!replicaToBeMoved.isLeader()) {
+      return true;
+    }
+    String topicName = replicaToBeMoved.topicPartition().topic();
+    if (!_mustHaveTopicLeadersPerBroker.contains(topicName)) {
+      return true;
+    }
+    // Moving a replica leadership from a not-alive broker to an alive broker is self-satisfied
+    if (!sourceBroker.isAlive() && destinationBroker.isAlive()) {
+      return true;
+    }
+    // Moving leader replica from more abundant broker could be considered as self-satisfied
+    return sourceBroker.numLeaderReplicasOfTopicInBroker(topicName) > destinationBroker.numLeaderReplicasOfTopicInBroker(topicName);
```

Comment:
Shouldn't this be 
```
return sourceBroker.numLeaderReplicasOfTopicInBroker(topicName) > minTopicLeadersPerBroker();
```

Otherwise your move can cause (1) the source broker to have fewer than intended number of replicas from the topic or (2) extra replica moves from source to other brokers and from other brokers to source to compensate for the missing topic replicas.

Am I missing something?


-- Also with this change and all the other changes proposed above, this function will have a single line of code, which is only called from `selfSatisfied`. Hence, it can be deleted and its content can be moved to `selfSatisfied`.

Reply:
Good point!